### PR TITLE
[Security] Deprecate "always authenticate" and "exception on no token"

### PR DESCRIPTION
**Pull request by wouterj**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | yes
| Tickets       | Ref #41613
| License       | MIT
| Doc PR        | n/a

### CODE REVIEW

File: src/Symfony/Component/Security/Core/Authorization/AuthorizationChecker.php [hunk 1]
```diff
@@ -34,6 +34,13 @@ class AuthorizationChecker implements AuthorizationCheckerInterface
 
     public function __construct(TokenStorageInterface $tokenStorage, AuthenticationManagerInterface $authenticationManager, AccessDecisionManagerInterface $accessDecisionManager, bool $alwaysAuthenticate = false, bool $exceptionOnNoToken = true)
     {
+        if (false !== $alwaysAuthenticate) {
+            trigger_deprecation('symfony/security-core', '5.4', 'Not setting the 4th argument of "%s" to "false" is deprecated.', __METHOD__);
+        }
+        if (false !== $exceptionOnNoToken) {
+            trigger_deprecation('symfony/security-core', '5.4', 'Not setting the 5th argument of "%s" to "false" is deprecated.', __METHOD__);
```
Comment (chalasr) [inline src/Symfony/Component/Security/Core/Authorization/AuthorizationChecker.php:41]: Why not deprecate setting them completely?

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php [hunk 2]
```diff
@@ -90,7 +90,10 @@ public function getConfigTreeBuilder()
                     ->defaultValue(SessionAuthenticationStrategy::MIGRATE)
                 ->end()
                 ->booleanNode('hide_user_not_found')->defaultTrue()->end()
-                ->booleanNode('always_authenticate_before_granting')->defaultFalse()->end()
+                ->booleanNode('always_authenticate_before_granting')
+                    ->defaultFalse()
+                    ->setDeprecated('symfony/security-bundle', '5.4', 'The "%node%" at path "%path%" is deprecated and will always be "false" in 6.0.')
```
Comment (chalasr) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: Can't it be removed in 6.0?

File: src/Symfony/Component/Security/Core/Authorization/AuthorizationChecker.php (hunk 1 above)
Comment (wouterj) [reply src/Symfony/Component/Security/Core/Authorization/AuthorizationChecker.php:41]: Not setting `$exceptionOnNoToken` means you'll fall back to the legacy behavior (it's an opt-in flag). So it must be set to `false`, but we can safely remove the argument in Symfony 6.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php (hunk 2 above)
Comment (wouterj) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php]: Thanks for spotting this confusing wording. Yes, it'll be removed in 6.0

File: src/Symfony/Component/Security/Core/Authorization/AuthorizationChecker.php (hunk 1 above)
Comment (chalasr) [reply src/Symfony/Component/Security/Core/Authorization/AuthorizationChecker.php:41]: Thank you.

File: src/Symfony/Component/Security/Core/Authorization/AuthorizationChecker.php [hunk 3]
```diff
@@ -34,6 +34,13 @@ class AuthorizationChecker implements AuthorizationCheckerInterface
 
     public function __construct(TokenStorageInterface $tokenStorage, AuthenticationManagerInterface $authenticationManager, AccessDecisionManagerInterface $accessDecisionManager, bool $alwaysAuthenticate = false, bool $exceptionOnNoToken = true)
     {
+        if (false !== $alwaysAuthenticate) {
+            trigger_deprecation('symfony/security-core', '5.4', 'Not setting the 4th argument of "%s" to "false" is deprecated.', __METHOD__);
```
Comment (chalasr) [inline src/Symfony/Component/Security/Core/Authorization/AuthorizationChecker.php:38]: (minor) we might prefer talking about arguments names over positions in deprecation messages, see e.g. https://github.com/symfony/symfony/commit/8e3058d95acc7d177516bfacce794f64695c2b0d

**Review by chalasr — APPROVED** [review]
👍 with minor comment on the wording

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @wouterj.

**Pull request merged**